libc: Fix time() and implement gmtime()

sam-itt opened the pull request.

sam-itt committed:
libc: Fix time()

The offset between win32 and unix epochs wasn't taken care of,
resulting in a wrong value. Added the correct offset from ntfs3g.

sam-itt committed:
libc: Implement gmtime()

Provide a working gmtime() implementation instead of the
NULL-returning pdclib stub.

JayFoxRox commented on functions/time/gmtime.c:
```diff
@@ -1,16 +1,111 @@
+ * Original Author: Adapted from tzcode maintained by Arthur David Olson.
+ *
+ *  This file is part of the Public Domain C Library (PDCLib).
+ *  Permission is granted to use, modify, and / or redistribute at will.
```
This section is no longer true; copying from newlib is also problematic as it conflicts with the pdclib license and it's not clearly marked as such.

The file in newlib is actually problematic in itself because it doesn't have explicit copyright headers, but mentions a couple of contributors and external sources. Strictly speaking we'd have to add https://www.sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=COPYING.NEWLIB and also ship it with each release of an nxdk app - this feels overkill to me.

I guess we could say that some of these algorithms are fact and non-copyrightable (as are the constants), but it still feels "bad".

sam-itt replied:
Well, there even is a pointer in the comments to a github page that describe the algorithm. Being what it is (converting seconds since 1970 to day/month/etc) I don't think there are so many ways to do it.

JayFoxRox commented on platform/xbox/functions/time/time.c:
```diff
@@ -9,13 +9,15 @@
 #ifndef REGTEST
 
 #include <xboxkrnl/xboxkrnl.h>
+#define NT_EPOCH_TIME_OFFSET ((LONGLONG)(369 * 365 + 89) * 24 * 3600 * 10000000)
```
I'm surprised this works (does it?). `3600 * 10000000` is more than 32-bit. I'd assume this has to be `3600LL * 10000000LL` etc. (I keep forgetting about C int promotion rules, so maybe it's already fine?)

sam-itt replied:
It works. That's from ntfs-3g ntfstime.h. The compiler uses the first type that fits. Better explained there: https://stackoverflow.com/questions/13134956/what-is-the-reason-for-explicitly-declaring-l-or-ul-for-long-values.

JayFoxRox commented on functions/time/gmtime.c:
```diff
@@ -1,16 +1,111 @@
+ * From newlib gmtime_r.c
+ * Original Author: Adapted from tzcode maintained by Arthur David Olson.
+ *
+ *  This file is part of the Public Domain C Library (PDCLib).
```
pdclib is designed in a way that all platform dependent code can be moved into the platform code. It has one-function per platform.

So either try to upstream this to https://github.com/DevSolar/pdclib in this folder, or move the file to platform/xbox/functions/time/gmtime.c

Going forward I assume we'll keep reverting the `functions/` folder to be vanilla upstream (I think the major roadblock is currently our Makefile, which probably uses wildcards).

sam-itt replied:
Done moving the file. I've also deleted the stub : Keeping it makes the final lib use it and thus return NULL which in turn crashes my app (that was the reason that led me doing this date-related stuff).

JayFoxRox commented on functions/time/gmtime.c:
```diff
@@ -1,16 +1,111 @@
+static struct tm res;
 
-struct tm * gmtime( const time_t * timer )
+struct tm * gmtime( const time_t * tim_p )
```
I assume this breaks consistency with the header. I'd recommend to keep the pdclib variant for consistency with other functions (I also assume it's closer to C spec?).

sam-itt replied:
Done

sam-itt committed:
xbox: Add Win32->Unix error translation function

sam-itt committed:
Move from OpenXDK X*File* to Win32 API

sam-itt committed:
Fix PDCLib fwrite segfault

Not resetting last seen '\n' offset after flushing the buffer
(which resets the buffer index to 0) can lead to a situation where
that offset has a higher value the the current buffer index,
overflowing (unsigned) stream->bufidx = bufidx - offset
and then doing a  memmove of a gigantic size.

JayFoxRox commented:
This includes the other changes now, none of which follow prefix style or have anything to do with this.
I think this might have been a mistake during a rebase?

---

Other than that I think the changes made the PR fine. I was hoping someone else would also review / merge this.

sam-itt commented:
> I think this might have been a mistake during a rebase?

I messed up. Here is a clean branch with only the intented changes: #24

JayFoxRox commented:
Oh! I see. Default branch for PR.

I'm closing this PR.
